File: internal/server/dto/gym.go
package dto

import "github.com/Topvennie/beta-log/internal/database/model"

type Gym struct {
	ID       int          `json:"id"`
	Name     string       `json:"name"`
	IconPath string       `json:"icon_path"`
	Source   model.Source `json:"source"`
}

func GymDTO(g *model.Gym) Gym {
	return Gym{
		ID:       g.ID,
		Name:     g.Name,
		IconPath: g.IconPath,
		Source:   g.Source,
	}
}

type GymCreate struct {
	Name     string `json:"name" validate:"required"`
	IconPath string `json:"icon_path"`
}

func (g GymCreate) ToModel() model.Gym {
	return model.Gym{
		Name:     g.Name,
		IconPath: g.IconPath,
	}
}

type GymUpdate struct {
	ID       int    `json:"id" validate:"required,min=1"`
	Name     string `json:"name" validate:"required"`
	IconPath string `json:"icon_path"`
}

func (g GymUpdate) ToModel() model.Gym {
	return model.Gym{
		ID:       g.ID,
		Name:     g.Name,
		IconPath: g.IconPath,
	}
}

type GymStatsVisits struct {
	Gym    string `json:"gym"`
	Amount int    `json:"amount"`
}

type GymStatsTop struct {
	Gym   string `json:"gym"`
	Top   int    `json:"top"`
	Flash int    `json:"flash"`
}

type GymStatsDistribution struct {
	Gym          string      `json:"gym"`
	Distribution map[int]int `json:"distribution"`
}

type GymStats struct {
	Total             int                    `json:"total"`
	MostVisited       string                 `json:"most_visited"`
	MostVisitedAmount int                    `json:"most_visited_amount"`
	Sessions          int                    `json:"sessions"`
	GraphVisits       []GymStatsVisits       `json:"graph_visits"`
	GraphTop          []GymStatsTop          `json:"graph_top"`
	GraphDistribution []GymStatsDistribution `json:"graph_distribution"`
}

File: ui/src/lib/types/gym.ts
import z from "zod";
import type { API } from "./api";
import { JSONBody } from "./general";

export enum Source {
  Toplogger = "toplogger",
  Manual = "manual",
}


export interface Gym {
  id: number;
  name: string;
  iconPath: string;
  source: Source;
}

// Converts

export const convertGym = (g: API.Gym): Gym => ({
  id: g.id,
  name: g.name,
  iconPath: g.icon_path,
  source: g.source as Source,
});

export const convertGyms = (g: API.Gym[]): Gym[] => g.map(convertGym);

export const convertGymUpdateSchema = (g: Gym): GymUpdate => {
  return {
    id: g.id,
    name: g.name,
    iconPath: g.iconPath,
  }
}

export interface GymStatsVisits {
  gym: string;
  amount: number;
}

export interface GymStatsTop {
  gym: string;
  top: number;
  flash: number;
}

export interface GymStatsDistribution {
  gym: string;
  distribution: Record<number, number>;
}

export interface GymStats {
  total: number;
  mostVisited: string;
  mostVisitedAmount: number;
  sessions: number;
  graphVisits: GymStatsVisits[];
  graphTop: GymStatsTop[];
  graphDistribution: GymStatsDistribution[];
}

export const convertGymStats = (s: API.GymStats): GymStats => ({
  total: s.total,
  mostVisited: s.most_visited,
  mostVisitedAmount: s.most_visited_amount,
  sessions: s.sessions,
  graphVisits: s.graph_visits,
  graphTop: s.graph_top,
  graphDistribution: s.graph_distribution,
});

// Schemas

export const gymCreateSchema = z.object({
  name: z.string().min(1),
  iconPath: z.string().optional(),
})
export type GymCreate = z.infer<typeof gymCreateSchema> & JSONBody

export const gymUpdateSchema = z.object({
  id: z.number().positive(),
  name: z.string().min(1),
  iconPath: z.string().optional(),
})
export type GymUpdate = z.infer<typeof gymUpdateSchema> & JSONBody
